File: frontend/src/game/game-ui.ts
import * as _ from 'lodash'
import {images} from '../common/images'
import {eventSequenceDuration, renderGame} from '../gameBoard/animated'
import {h} from 'snabbdom'
import {numberOfInstructionsPerCycle} from '../common/constants'
import {GameState} from "../state";

export function render(state: GameState, actionHandler) {
  return h('div.game', [
    fab('.fab-right-1', images.iconClose, [actionHandler, {leaveGame: true}]),
    fab('.fab-left-1', images.iconReplayAnimation, event => {
      const svg = document.querySelector('.game-board svg')
      if (svg)
        (<HTMLMediaElement>svg).currentTime = eventSequenceDuration(
          _.takeWhile(state.game.events, (event: any) => !event.StartCycleEvaluation || event.StartCycleEvaluation.cycle !== state.game.cycle - 1)
        )
    }),
    renderGame(state.game),
    renderActionButtons(state, actionHandler)
  ])
}

function fab(classes, image, onclick) {
  return h('div.fab' + classes, {on: {click: onclick}},
    h('img', {props: {src: image}}))
}

function robotImage(index, filled, you, onclick) {
  return h('img', {
    class: {
      'robot-tile': true,
      'robot-tile-you': you
    },
    props: {
      src: filled ? images.player(index) : images.playerStart(index)
    },
    on: {click: onclick}
  })
}

function renderActionButtons(state: GameState, actionHandler) {
  const focusedSlot = state.focusedSlot || 0
  const player: any = _.get(state, 'game.you')

  function instructionCard(type) {
    function unusedAndThisType(opt, index) {
      return opt[type] && !_.some(player.instructionSlots, slot => slot === index)
    }

    const image = images.action(type)
    const count = player.instructionOptions.filter(unusedAndThisType).length
    const unusedIndex = _.findIndex(player.instructionOptions, unusedAndThisType)
    return h('div.action',
      {
        on: {
          click: count !== 0 ? () => actionHandler({
            setInstruction: true,
            slot: focusedSlot,
            instruction: unusedIndex
          }) : undefined
        }
      },
      [h('img', {props: {src: image}}), h('div.badge', count)])
  }

  function instructionSlot(index) {
    const instruction = player.instructionSlots[index]
    const focused = focusedSlot === index
    const props = {
      class: {focused},
      on: {
        dblclick: () => actionHandler({resetSlot: true, slot: index}),
        click: () => actionHandler({focusSlot: index})
      }
    }
    if (instruction !== undefined && instruction !== null) {
      const image = images.action(Object.keys(player.instructionOptions[instruction])[0])
      return h('span.slot.filled', props,
        h('img', {props: {src: image}}))
    } else {
      return h('span.slot', props, index + 1)
    }
  }

  if (!player && state.game.cycle === 0) {
    return h('div.text-panel', state.game.scenario.initialRobots
      .map(robot => robotImage(robot.index, false, false, () => actionHandler({joinGame: robot.index}))))
  } else if (!player) {
    return h('div.text-panel', 'observer mode')
  } else if (player.finished && player.finished.rageQuitted === false) {
    return h('div.text-panel', 'target reached as ' + player.finished.rank)
  } else if (player.finished && player.finished.rageQuitted === true) {
    return h('div.text-panel', 'game quitted')
  } else {
    let instructionTypes = []
    let instr = _.clone(player.instructionOptions)
    while (instr.length !== 0) {
      let head = instr[0]
      let type = Object.keys(head)[0]
      instructionTypes.push(type)
      instr = _.dropWhile(instr, i => i[type])
    }

    return h('div.footer-group', [
      h('div.slots-panel', _.range(numberOfInstructionsPerCycle).map(instructionSlot)),
      h('div.cards-panel', instructionTypes.map(instructionCard))
    ])
  }
}

File: frontend/src/frontend/game/GameUi.scala
package frontend.game

import com.raquo.snabbdom.simple.VNode
import com.raquo.snabbdom.simple.events.{onClick, onDblClick}
import com.raquo.snabbdom.simple.implicits._
import com.raquo.snabbdom.simple.props.{className, src}
import com.raquo.snabbdom.simple.attrs.id
import com.raquo.snabbdom.simple.tags._
import frontend.Main
import frontend.common.{Fab, Images}
import frontend.gameBoard.{Animation, RenderGame}
import frontend.util.{Dynamic, Ui}
import gameEntities._
import org.scalajs.dom

import scala.concurrent.ExecutionContext.Implicits.global

object GameUi extends Ui {
  def apply(gameState: GameState, rerender: GameState => Unit): VNode = {
    val replayFab = Fab("fab-left-1", Images.iconReplayAnimation, () => {
      Dynamic(dom.document.querySelector(".game-board svg")).setCurrentTime(
        if (gameState.game.cycle == 0)
          0
        else
          Animation.eventSequenceDuration(gameState.game.events.takeWhile {
            case s: StartCycleEvaluation => s.cycle != gameState.game.cycle - 1
            case _                       => true
          })
      )
    })
    val returnToLobbyFab = Fab("fab-right-1", Images.iconClose, () => Main.gotoLobby())

    gameState.game.you match {
      case None if gameState.game.cycle == 0 =>
        div(id := "robo-race",
          className := "game",
          returnToLobbyFab,
          replayFab,
          RenderGame(gameState.game, Some { (p, _) =>
            gameState.game.scenario.initialRobots.find(r => r.position == p && !gameState.game.robots.contains(r))
              .foreach(r => SendCommand(gameState, RegisterForGame(r.index))
                .foreach(rerender))
          }),
          div(className := "text-panel", "tap a start position to join the game")
        )

      case None =>
        div(id := "robo-race",
          className := "game",
          returnToLobbyFab,
          replayFab,
          RenderGame(gameState.game, None),
          div(className := "text-panel", "observer mode")
        )

      case Some(you) if you.finished.exists(_.rageQuitted == true) =>
        div(id := "robo-race",
          className := "game",
          returnToLobbyFab,
          replayFab,
          RenderGame(gameState.game, None),
          div(className := "text-panel", "game quitted")
        )

      case Some(you) if you.finished.exists(_.rageQuitted == false) =>
        div(id := "robo-race",
          className := "game",
          returnToLobbyFab,
          replayFab,
          RenderGame(gameState.game, None),
          div(className := "text-panel", "target reached as " + you.finished.get.rank)
        )

      case Some(you) =>
        div(id := "robo-race",
          className := "game",
          Fab("fab-right-1", Images.iconClose, () => SendCommand(gameState, DeregisterForGame).foreach(rerender)),
          replayFab,
          RenderGame(gameState.game, None),
          renderInstructionBar(gameState, rerender, you)
        )
    }
  }

  def renderInstructionBar(state: GameState, rerender: GameState => Unit, player: Player): VNode = {
    def instructionSlot(index: Int): VNode = {
      val instruction = player.instructionSlots(index)
        .map(player.instructionOptions.apply)
      val focused = state.focusedSlot == index

      val setFocus = onClick := (_ => rerender(state.copy(focusedSlot = index)))
      val resetSlot = onDblClick := (_ => SendCommand(state,
        ResetInstruction(
          cycle = state.game.cycle,
          slot = index))
        .foreach(rerender))

      instruction match {
        case Some(instr) =>
          span(className := "slot filled" + (if (focused) " focused" else ""),
            setFocus,
            resetSlot,
            img(src := Images.action(instr)),
          )
        case None        =>
          span(className := "slot" + (if (focused) " focused" else ""),
            setFocus,
            resetSlot,
            (index + 1).toString
          )
      }
    }

    def instructionCard(instruction: Instruction): VNode = {
      val ofThisTypeAndUnused = for {
        (instr, index) <- player.instructionOptions.zipWithIndex
        if instr == instruction
        if !player.instructionSlots.exists(_.contains(index))
      } yield index

      div(className := s"action stacked-action-${ofThisTypeAndUnused.size.min(5)}",
        img(src := Images.action(instruction)),
        if (ofThisTypeAndUnused.nonEmpty) div(className := "badge", ofThisTypeAndUnused.size.toString)
        else None,
        ofThisTypeAndUnused.headOption match {
          case Some(index) => onClick := (_ => SendCommand(state,
            SetInstruction(
              cycle = state.game.cycle,
              slot = state.focusedSlot,
              instruction = index))
            .map(selectNextFreeSlot(state))
            .foreach(rerender))
          case None        => None
        }
      )
    }

    div(className := "footer-group",
      div(className := "slots-panel",
        seq((0 until Constants.instructionsPerCycle).map(instructionSlot))),
      div(className := "cards-panel",
        seq(player.instructionOptions.distinct.map(instructionCard)))
    )
  }

  def selectNextFreeSlot(oldState: GameState)(gameState: GameState): GameState =
    if (gameState.game.you.isDefined) {
      val slot = if (oldState.game.cycle != gameState.game.cycle)
        0
      else
        (for {
          i <- 0 until Constants.instructionsPerCycle
          s = (i + gameState.focusedSlot) % Constants.instructionsPerCycle
          if gameState.game.you.get.instructionSlots(s).isEmpty
        } yield s).headOption.getOrElse(0)
      gameState.copy(focusedSlot = slot)
    } else {
      gameState
    }
}
